gift-certificate-application/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -200,35 +206,78 @@ public List<GiftCertificate> findByAttributes(String name, List<String> tagNames
         checkColumnNames(columnNames);
         checkSortType(sortType);
 
-        if (!paginationValidator.isSizeValid(pagination)) {
+        if (pagination.getPage() == ZERO && pagination.getSize() == ZERO) {
+            throw new PaginationException(Message.CHOOSE_PAGINATION_MSG);
+        }
+
+        Pageable pageable;
+        if (columnNames != null) {
+            Sort sort = createSort(columnNames, sortType);
+            pageable = createPageableWithSort(pagination, sort);
+        } else {
+            pageable = paginationConverter.convert(pagination);
+        }
+
+        Page<GiftCertificate> giftCertificatePage;
+        if (name == null && tagNames ==null) {
+            giftCertificatePage = giftCertificateRepository.findAll(pageable);
+        } else {
+            giftCertificatePage = giftCertificateRepository.findByAttributes(name, tagNames, pageable);
+        }
+
+        int lastPage = giftCertificatePage.getTotalPages();
+        if(!paginationValidator.isPageValid(pagination, lastPage)) {
+            String message = String.format(Message.PAGE_NUMBER_INVALID_MSG, lastPage);
+            throw new PageNumberValidationException(message);
+        }
+
```

Comment:
please separate this method into several private methods, like "preparePageable" and so on, to make it easy to read